packages/amplify-provider-awscloudformation/src/cloud-formation-error-handler.ts

```diff
@@ -0,0 +1,34 @@
+import { $TSAny, AmplifyError } from 'amplify-cli-core';
+
+export const handleCloudFormationError = (err: Error & { details?: string }): void => {
+  if (err?.name === 'ValidationError' && err?.message === 'No updates are to be performed.') {
+    return;
+  }
+
+  if (err?.name === 'ValidationError' && (err?.message ?? '').includes('_IN_PROGRESS state and can not be updated.')) {
+    throw new AmplifyError(
+      'DeploymentInProgressError',
+      {
+        message: 'Deployment is already in progress.',
+        resolution: 'Wait for the other deployment to finish and try again.',
+        code: (err as $TSAny).code,
+      },
+      err,
+    );
+  }
+
+  const s3Indicator = '(AWS::S3::Bucket)';
+  const alreadyExistsSuffix = 'already exists';
+  if (err?.details?.includes(s3Indicator) && err?.details?.includes(alreadyExistsSuffix)) {
+    const prefix = `Name: ${
+      err.details.includes('CustomMessageConfirmationBucket') ? 'CustomMessageConfirmationBucket' : 'S3Bucket'
+    } (AWS::S3::Bucket), Event Type: create, Reason:`;
+    const bucketName = err.details.slice(prefix.length + 1, err.details.length - alreadyExistsSuffix.length - 1);
```

Comment:
1. This won't work in few cases since `err.details` is a `\n` separated string of all CFN errors. The s3 bucket error may or may not be the first one. e.g a sample `err.details` 
```
Name: <stackName> (AWS::CloudFormation::Stack), Event Type: update, Reason: Parameters: [a1] must have values
Name: S3Bucket (AWS::S3::Bucket), Event Type: create, Reason: <bucketName> already exists
```
2. There are errors with S3 bucket's resource name as DeploymentBucket e.g. `Name: DeploymentBucket (AWS::S3::Bucket), Event Type: create, Reason: <bucketName> already exists`. Is the root cause not the same for this occurrence?

We should look into adding a type for serializing and deserializing the CFN errors. The serialization happens [here](https://github.com/aws-amplify/amplify-cli/blob/137b629c9e74e4ab8ec90f4519f947ed748f47cf/packages/amplify-provider-awscloudformation/src/aws-utils/aws-cfn.js#L145-L155)

Reply:
This is an interesting idea, maybe we can deserialize it into something that is more predictable and consistent?

Reply:
1. Thank you for catching this. I missed the possibility of multi-line details fields. I just updated the PR to first find the line that contains this error and then parse the bucket name from just this line.
2. I was not able to reproduce the DeploymentBucket error yet. Though if it occurs when pulling an existing project, I do not think that it is the same root cause.

Reply:
I implemented functions to serialize/deserialize, as well as types to hold the deserialized data.